src/foam/u2/view/RichChoiceView.js

```diff
@@ -267,16 +279,18 @@ foam.CLASS({
       this.onDataUpdate();
 
       this
-        .attrs({ name: this.name })
+        .attrs({
+          name: this.name,
+          tabIndex: 0
+        })
         .addClass(this.myClass())
+        .on('keydown', function(evt) {
+          if ( evt.which === 13 || evt.keyCode === 13 ) self.toggle();
```

Comment:
Mozilla Developer Network now recommends using evt.key to check the key.

This would be a good time to write a utility or polyfill so we only need to call a single function to check key sizes.

Maybe foam.web.key(evt) which can check evt.key evt.keyCode evt.charCode evt.which, etc. Then our keyboard event handling will be shorter and more cross platform everywhere we do things like this :)

Reply:
I'm with you on the spirit of the idea, but after looking into it I don't think it's going to work.

`key` is preferred and `which`, `keyCode`, and `keyIdentifier` are all deprecated. The problem is that the types of those properties vary, so I'm not sure what the method signature would look like. `key` is a string like "Enter" and `which` and `keyCode` are integers like 13. So would our utility method take one of the two (ie: either the string or the int) and then have to use a huge table to convert between them? I don't particularly like the idea of creating that table.

An alternative is to allow them to supply either the string or the int, but that's obviously not a great plan since they have to be aware of the implementation at that point practically anyway.

And finally if we stick to just one (string, let's say) and don't make the lookup table, then it's no different from the caller just checking `key` directly.

On top of all of that, the developer has to know to use the utility in the first place.

So I'm struggling to see how we can make a useful utility for this case.

Reply:
The utility method would take the keyboard event and return the key identifier for that event.  It would just use .key if it exists, or it would interpret keyCode/charCode/which as needed.

A table to convert from deprecated keyCode/charCode/which to the non-deprecated "key" identifier would be an entirely reasonable thing to have.  I agree it wouldn't be fun to create, but solving it now prevents us having to play whack a mole to find these bugs later or crossing our fingers hoping that we catch them all in review.  For somebody less familiar with the situation, reading code that checks "keyCode" "which" or any of the existing properties on the event seems entirely reasonably, its only when you have a reviewer with really specific domain knowledge that you will catch it.

It might not even require a very big table.  Most characters return their printed representation, it's only control and unprintable characters that require mapping to some known name.  So through some clever combination of String.fromCodePoint and checking the range that the key code lies in might not be that much work.

The developer has to know to use the utility but that can be said of literally any feature we provide.  I would rather solve this problem than train every developer to recognize this edge case of web programming.  Even to many experienced web programmers, code that checks "keyCode" looks entirely reasonable.  Additionally we can take steps to force or encourage the developer to use our interface.  If we decorate the listener during .on() in foam.u2.Element, we don't have to give them the raw keyboard event.  We could instead return them our own custom modeled keyboard event.  Or we could do other fun things like modify the KeyboardEvent.prototype to remove keyCode/charCode when running in developer mode, preventing any FOAM dev from getting useful results from those properties.

A lot of the FOAM framework (and any web framework) is providing better and more consistent interfaces to the browser than are provided out of box.  It's part of why we have the u2 package in the first place.  The web is a strange and messy platform. The less domain specific knowledge people need to build apps in FOAM, the fewer bugs they will create and the more productive they will be.

Reply:
@adamvy Updated to include the utility. Please review again.